### ISSUE TITLE
[BUG] Models using nn.Sequential may fail to train following PR #1764

### ISSUE THREAD

**Issue by drckf** [ml-explore/mlx#1853]
**Describe the bug**

Models using nn.Sequential fail to train following PR #1764 (Limit grad recursion depth by not recursing through non-grad inputs). The issue is reproducible with a minimal example that demonstrates identical model architectures failing or succeeding based solely on whether nn.Sequential is used. Note that I did my best to isolate the issue to a specific PR but, of course, I could be wrong. 

Key observations:
- Models using nn.Sequential fail to train, staying at 0.089 accuracy
- Identical models using direct layer iteration train successfully, reaching 0.909 accuracy
- Issue is reproducible across different datasets (MNIST, FashionMNIST)
- The only difference between working/non-working implementations is the use of Sequential:

```python
# Fails to train (stays at 0.089 accuracy)
self.model = nn.Sequential(*self.layers)
def __call__(self, x):
    return self.model(x)

# Trains successfully (reaches 0.909 accuracy)
def __call__(self, x):
    for layer in self.layers:   
        x = layer(x)
    return x
```

The issue appears to be related to the gradient recursion depth changes introduced in PR #1764. That is, both versions of the model train to 91% accuracy on PR #1763, but only the one without Sequential trains on all subsequent PRs. 

Below are some test results demonstrating the regression across different versions of mlx.

**To Reproduce**

Include code snippet
```python


from functools import partial

import mlx.core as mx
import mlx.nn as nn
import mlx.data as dx
import mlx.optimizers as opt
import numpy as np

import supermnist


def seed_everything(seed):
    mx.random.seed(seed)
    np.random.seed(seed)


def default_collate(batch):
    keys = batch[0].keys()
    collated = {}
    for k in keys:
        values = [sample[k] for sample in batch]
        collated[k] = np.stack(values)
    return collated


class SimpleDataset:
    def __init__(self, data, keys, transform= None):
        self.buffer = dx.buffer_from_vector(data)
        self.keys = keys
        self.transform = transform
        
    def __len__(self):
        return len(self.buffer)
    
    def __getitem__(self, idx):
        sample = dict(self.buffer[idx])  
        sample = self.transform_and_filter(sample)
        return sample

    def __iter__(self):
        for i in range(len(self)):
            yield self[i]
    
    def transform_and_filter(self, sample):
        if self.transform is not None:
            sample = self.transform(sample)
        if self.keys is not None:
            sample = {k: sample[k] for k in self.keys}
        return sample


class SimpleDataLoader:
    def __init__(self, dataset, batch_size=256, shuffle=True):
        self.dataset = dataset
        self.batch_size = batch_size
        self.shuffle = shuffle

    def __iter__(self):
        indices = list(range(len(self.dataset)))
        if self.shuffle:
            np.random.shuffle(indices)

        # drop the last batch if it's not full
        truncated = len(indices) - (len(indices) % self.batch_size)
        indices = indices[:truncated]

        for i in range(0, len(indices), self.batch_size):
            batch_indices = indices[i:i + self.batch_size]
            batch_samples = [self.dataset[idx] for idx in batch_indices]
            yield default_collate(batch_samples)

    def __len__(self):
        return len(self.dataset) // self.batch_size


def loss_fn(model, X, y):
    return nn.losses.cross_entropy(model(X), y, reduction="mean")
    

class BadMLP(nn.Module):
    def __init__(
            self, 
            num_layers: int, 
            input_dim: int, 
            hidden_dim: int, 
            output_dim: int
        ):
        super().__init__()
        self.layers = [
            nn.Linear(input_dim, hidden_dim),
            nn.ReLU()
        ]
        for _ in range(num_layers - 1):
            self.layers += [
                nn.Linear(hidden_dim, hidden_dim),
                nn.ReLU()
            ]
        self.layers += [
            nn.Linear(hidden_dim, output_dim)
        ]
        self.model = nn.Sequential(*self.layers)

    def __call__(self, x):
        return self.model(x)
    

class GoodMLP(nn.Module):
    def __init__(
            self, 
            num_layers: int, 
            input_dim: int, 
            hidden_dim: int, 
            output_dim: int
        ):
        super().__init__()
        self.layers = [
            nn.Linear(input_dim, hidden_dim),
            nn.ReLU()
        ]
        for _ in range(num_layers - 1):
            self.layers += [
                nn.Linear(hidden_dim, hidden_dim),
                nn.ReLU()
            ]
        self.layers += [
            nn.Linear(hidden_dim, output_dim)
        ]

    def __call__(self, x):
        for layer in self.layers:   
            x = layer(x)
        return x
        

def main(
        dataset="digits", 
        model_type="bad",
        seed = 0,
        num_layers = 2,
        hidden_dim = 64,
        num_classes = 10,
        batch_size = 256,
        num_epochs = 20,
        learning_rate = 1e-2
    ):

    seed_everything(seed)
    print(mx.__version__)

    # Load the training data
    train_images, train_labels = supermnist.load(dataset=dataset, split="train")
    train_images = train_images.astype(np.float32) / 255.0
    train_data = [
        {"image": img, "label": lbl} for img, lbl in zip(train_images, train_labels)
    ]
    train_dataset = SimpleDataset(train_data, keys=("image", "label"))
    train_loader = SimpleDataLoader(
        dataset=train_dataset,
        batch_size=batch_size,
        shuffle=True
    )

    # Load the test data
    test_images, test_labels = supermnist.load(dataset=dataset, split="test")
    test_images = test_images.astype(np.float32) / 255.0
    test_images = mx.array(test_images)
    test_labels = mx.array(test_labels)

    # Create the model
    if model_type == "bad":
        model = BadMLP(
            num_layers=num_layers, 
            input_dim=train_images.shape[-1], 
            hidden_dim=hidden_dim, 
            output_dim=num_classes
        )
    elif model_type == "good":
        model = GoodMLP(
            num_layers=num_layers, 
            input_dim=train_images.shape[-1], 
            hidden_dim=hidden_dim, 
            output_dim=num_classes
        )
    mx.eval(model.parameters())

    # Create the optimizer
    optimizer = opt.SGD(learning_rate=learning_rate)

    # Create the loss and gradient function
    loss_and_grad_fn = nn.value_and_grad(model, loss_fn)

    # Compile the step function
    @partial(mx.compile, inputs=model.state, outputs=model.state)
    def step(X, y):
        loss, grads = loss_and_grad_fn(model, X, y)
        optimizer.update(model, grads)
        return loss

    # Compile the evaluation function
    @partial(mx.compile, inputs=model.state)
    def eval_fn(X, y):
        return mx.mean(mx.argmax(model(X), axis=1) == y)

    # Train the model
    for e in range(num_epochs):
        for batch in train_loader:
            step(
                mx.array(batch["image"]), 
                mx.array(batch["label"])
            )
            mx.eval(model.state)
        accuracy = eval_fn(test_images, test_labels)
        print(
            f"Epoch {e}: Test accuracy {accuracy.item():.3f},"
        )


if __name__ == "__main__":
    mx.set_default_device(mx.cpu)
    main(model_type="bad")
```

With conda environment:

```yaml
# Machine Learning Conda Environment

name: ml-research
channels:
  - pytorch
  - conda-forge
  - defaults
dependencies:
  # Core Scientific Python
  - python=3.12
  - numpy<2.0
  - pandas
  - scipy
  - scikit-learn
  
  # Deep Learning
  - pytorch
  - torchvision
  - torchaudio
  
  # Distributed Computing
  - ray-default
  - dask
  
  # Experiment Tracking
  - mlflow
  
  # Visualization
  - matplotlib
  - seaborn
  - plotly
  
  # Jupyter Environment
  - jupyter
  - jupyterlab
  - ipywidgets
  
  # Data Processing
  - pyarrow
  - polars
  
  # Model Evaluation & Training
  - optuna
  
  # API and Data Validation
  - fastapi
  - uvicorn
  - pydantic
  
  # Utilities
  - tqdm
  - pytest
  - black
  - flake8
  - isort
  
  # Install additional packages using pip
  - pip
  - pip:
    - accelerate
    - mlx==0.22.1 # note that we need to build from source for debug specific verisons
    - mlx-data==0.0.2
    - git+ssh://git@github.com/bad-country/SuperMNIST.git
```

**Expected behavior**
This is a simple MLP classifier trained on MNIST. Prior to PR #1764 both versions of the model (with or without Sequential) train to about 91% accuracy in 20 epochs and training dynamics are identical. After PR #1764, the model without Sequential is unaffected but the model with Sequential fails to train. Outputs for various mlx versions are as follows:

```
# PR 1763:

## model_type=”good” (without Sequential)

Output:
0.22.0.dev20250210+a4a2764a
Epoch 0: Test accuracy 0.452,
Epoch 1: Test accuracy 0.559,
Epoch 2: Test accuracy 0.713,
Epoch 3: Test accuracy 0.794,
Epoch 4: Test accuracy 0.827,
Epoch 5: Test accuracy 0.852,
Epoch 6: Test accuracy 0.866,
Epoch 7: Test accuracy 0.876,
Epoch 8: Test accuracy 0.880,
Epoch 9: Test accuracy 0.887,
Epoch 10: Test accuracy 0.892,
Epoch 11: Test accuracy 0.895,
Epoch 12: Test accuracy 0.898,
Epoch 13: Test accuracy 0.900,
Epoch 14: Test accuracy 0.902,
Epoch 15: Test accuracy 0.903,
Epoch 16: Test accuracy 0.904,
Epoch 17: Test accuracy 0.907,
Epoch 18: Test accuracy 0.909,
Epoch 19: Test accuracy 0.909,

## model_type=”bad” (with Sequential)

Output:
0.22.0.dev20250210+a4a2764a
Epoch 0: Test accuracy 0.452,
Epoch 1: Test accuracy 0.559,
Epoch 2: Test accuracy 0.713,
Epoch 3: Test accuracy 0.794,
Epoch 4: Test accuracy 0.827,
Epoch 5: Test accuracy 0.852,
Epoch 6: Test accuracy 0.866,
Epoch 7: Test accuracy 0.876,
Epoch 8: Test accuracy 0.880,
Epoch 9: Test accuracy 0.887,
Epoch 10: Test accuracy 0.892,
Epoch 11: Test accuracy 0.895,
Epoch 12: Test accuracy 0.898,
Epoch 13: Test accuracy 0.900,
Epoch 14: Test accuracy 0.902,
Epoch 15: Test accuracy 0.903,
Epoch 16: Test accuracy 0.904,
Epoch 17: Test accuracy 0.907,
Epoch 18: Test accuracy 0.909,
Epoch 19: Test accuracy 0.909,

# PR 1764:

## model_type=”good” (with Sequential)

Output:
0.22.0.dev20250210+33421c1d
Epoch 0: Test accuracy 0.452,
Epoch 1: Test accuracy 0.559,
Epoch 2: Test accuracy 0.713,
Epoch 3: Test accuracy 0.794,
Epoch 4: Test accuracy 0.827,
Epoch 5: Test accuracy 0.852,
Epoch 6: Test accuracy 0.866,
Epoch 7: Test accuracy 0.876,
Epoch 8: Test accuracy 0.880,
Epoch 9: Test accuracy 0.887,
Epoch 10: Test accuracy 0.892,
Epoch 11: Test accuracy 0.895,
Epoch 12: Test accuracy 0.898,
Epoch 13: Test accuracy 0.900,
Epoch 14: Test accuracy 0.902,
Epoch 15: Test accuracy 0.903,
Epoch 16: Test accuracy 0.904,
Epoch 17: Test accuracy 0.907,
Epoch 18: Test accuracy 0.909,
Epoch 19: Test accuracy 0.909,

## model_type=”bad” (without Sequential)

Output:
0.22.0.dev20250210+33421c1d
Epoch 0: Test accuracy 0.089,
Epoch 1: Test accuracy 0.089,
Epoch 2: Test accuracy 0.089,
Epoch 3: Test accuracy 0.089,
Epoch 4: Test accuracy 0.089,
Epoch 5: Test accuracy 0.089,
Epoch 6: Test accuracy 0.089,
Epoch 7: Test accuracy 0.089,
Epoch 8: Test accuracy 0.089,
Epoch 9: Test accuracy 0.089,
Epoch 10: Test accuracy 0.089,
Epoch 11: Test accuracy 0.089,
Epoch 12: Test accuracy 0.089,
Epoch 13: Test accuracy 0.089,
Epoch 14: Test accuracy 0.089,
Epoch 15: Test accuracy 0.089,
Epoch 16: Test accuracy 0.089,
Epoch 17: Test accuracy 0.089,
Epoch 18: Test accuracy 0.089,
Epoch 19: Test accuracy 0.089,

# Release 0.22.1

## model_type=”good” (without Sequential)

Output:
0.22.1
Epoch 0: Test accuracy 0.452,
Epoch 1: Test accuracy 0.559,
Epoch 2: Test accuracy 0.713,
Epoch 3: Test accuracy 0.794,
Epoch 4: Test accuracy 0.827,
Epoch 5: Test accuracy 0.852,
Epoch 6: Test accuracy 0.866,
Epoch 7: Test accuracy 0.876,
Epoch 8: Test accuracy 0.880,
Epoch 9: Test accuracy 0.887,
Epoch 10: Test accuracy 0.892,
Epoch 11: Test accuracy 0.895,
Epoch 12: Test accuracy 0.898,
Epoch 13: Test accuracy 0.900,
Epoch 14: Test accuracy 0.902,
Epoch 15: Test accuracy 0.903,
Epoch 16: Test accuracy 0.904,
Epoch 17: Test accuracy 0.907,
Epoch 18: Test accuracy 0.909,
Epoch 19: Test accuracy 0.909,

## model_type=”bad” (with Sequential)

Output: 
0.22.1
Epoch 0: Test accuracy 0.089,
Epoch 1: Test accuracy 0.089,
Epoch 2: Test accuracy 0.089,
Epoch 3: Test accuracy 0.089,
Epoch 4: Test accuracy 0.089,
Epoch 5: Test accuracy 0.089,
Epoch 6: Test accuracy 0.089,
Epoch 7: Test accuracy 0.089,
Epoch 8: Test accuracy 0.089,
Epoch 9: Test accuracy 0.089,
Epoch 10: Test accuracy 0.089,
Epoch 11: Test accuracy 0.089,
Epoch 12: Test accuracy 0.089,
Epoch 13: Test accuracy 0.089,
Epoch 14: Test accuracy 0.089,
Epoch 15: Test accuracy 0.089,
Epoch 16: Test accuracy 0.089,
Epoch 17: Test accuracy 0.089,
Epoch 18: Test accuracy 0.089,
Epoch 19: Test accuracy 0.089,
```  

**Desktop (please complete the following information):**
-  M3 Macbook Air
 - OS Version: Sonoma 14.6.1

**Comment by awni** [comment]
Confirmed it is broken. Will investigate.

**Comment by angeloskath** [comment]
This should probably still work but the issue is that there are 2 ways to get the layers. If you simply change the bad MLP to the following (note the `self` missing from `layers`) it should work fine.

```python
class BadMLP(nn.Module):
    def __init__(
            self, 
            num_layers: int, 
            input_dim: int, 
            hidden_dim: int, 
            output_dim: int
        ):
        super().__init__()
        layers = [
            nn.Linear(input_dim, hidden_dim),
            nn.ReLU()
        ]
        for _ in range(num_layers - 1):
            layers += [
                nn.Linear(hidden_dim, hidden_dim),
                nn.ReLU()
            ]
        self.layers += [
            nn.Linear(hidden_dim, output_dim)
        ]
        self.model = nn.Sequential(*layers)

    def __call__(self, x):
        return self.model(x)
```

**Comment by drckf** [comment]
Yes, it has something to do with multiple references. For example, the following module doesn't train

```
class AnotherBadMLP(nn.Module):
    def __init__(
            self, 
            num_layers: int, # disregarded
            input_dim: int, 
            hidden_dim: int, 
            output_dim: int
        ):
        super().__init__()
        self.layer_1 = nn.Linear(input_dim, hidden_dim)
        self.activation = nn.ReLU()
        self.layer_2 = nn.Linear(hidden_dim, output_dim)
        self.layers = [self.layer_1, self.activation, self.layer_2]

    def __call__(self, x):
        x = self.layer_1(x)
        x = self.activation(x)
        x = self.layer_2(x)
        return x
```

but it trains fine if you comment out the `self.layers` definition. 

```
class AnotherBadMLP(nn.Module):
    def __init__(
            self, 
            num_layers: int, # disregarded
            input_dim: int, 
            hidden_dim: int, 
            output_dim: int
        ):
        super().__init__()
        self.layer_1 = nn.Linear(input_dim, hidden_dim)
        self.activation = nn.ReLU()
        self.layer_2 = nn.Linear(hidden_dim, output_dim)
        # self.layers = [self.layer_1, self.activation, self.layer_2]

    def __call__(self, x):
        x = self.layer_1(x)
        x = self.activation(x)
        x = self.layer_2(x)
        return x
```
